### PR TITLE
Add basic string helpers

### PR DESCRIPTION
**Pull request by sanjeevkpandit**
- Closes https://github.com/leapfrogtechnology/just-handlebars-helpers/issues/10
- Add basic string operators
 - `concat`
 - `join`
 - `lowercase`
 - `uppercase`
 - `first`
 - `last`

### CODE REVIEW

File: dist/h.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
...
+(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.H = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
+(function (global){
```
Comment (kabirbaidhya) [inline dist/h.js]: remove the changes to this `dist/*`

File: tests/helpers/strings.spec.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 
-import Handlebars from 'handlebars';
 import strings from '../../src/helpers/strings';
+import Handlebars from 'handlebars';
```
Comment (kabirbaidhya) [inline tests/helpers/strings.spec.js]: `import {compile} from 'handlebars';` as all we need to do is just `compile` the templates. 

And you could do just `compile(html)` instead of doing `Handlebars.compile(html)`.

File: src/helpers/strings.js [hunk 3]
```diff
@@ -1,5 +1,7 @@
-
-import { isFunction, isObject } from '../util/utils';
+import {
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: single line is better

File: src/helpers/strings.js [hunk 4]
```diff
@@ -1,4 +1,3 @@
-
 import { isFunction, isObject } from '../util/utils';
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: `import { isFunction, isObject, isString } from '../util/utils';`

File: src/helpers/strings.js [hunk 5]
```diff
@@ -120,5 +119,125 @@ export default {
...
+     */
+    lowercase: (param) => {
+        if (typeof param === 'undefined' || param === null) {
+            return '';
+        }
+
+        if (typeof param === 'string') {
+            return param.toLowerCase();
+        }
+
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: Don't need all these, just this would be fine: 

```javascript
lowercase: (param) => {
    return isString(param) ? param.toLowerCase() : param;
}
```

File: src/helpers/strings.js [hunk 6]
```diff
@@ -120,5 +119,125 @@ export default {
...
+    uppercase: (param) => {
+        if (typeof param === 'undefined' || param === null) {
+            return '';
+        }
+
+        if (typeof param === 'string') {
+            return param.toUpperCase();
+        }
+
+        return param;
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: same here

File: src/helpers/strings.js [hunk 7]
```diff
@@ -120,5 +119,125 @@ export default {
...
+    /**
+     * Get the first element of a collection/array.
+     * Example usage:
+     * 		{{first ['David', 'Miller', 'Jones']}}   => 'David'
+     *
+     * @param  array collection
+     * @return string
+     */
+    first: (collection) => {
+        if (typeof collection === 'undefined' || collection === null || collection.length === 0) {
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]:
```javascript
if (!isArray(collection) || collection.length ===0 ) {
   return '';
}}
```

File: src/helpers/strings.js [hunk 8]
```diff
@@ -120,5 +119,125 @@ export default {
...
+    /**
+     * Get the last element of a collection/array.
+     * Example usage:
+     * 		{{last ['David', 'Miller', 'Jones']}}   => 'Jones'
+     *
+     * @param  array collection
+     * @return string
+     */
+    last: (collection) => {
+        if (typeof collection === 'undefined' || collection === null || collection.length === 0) {
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: same here

File: src/helpers/strings.js (hunk 7 above)
Comment (kabirbaidhya) [inline src/helpers/strings.js]: need to import `{isArray}` as well

File: tests/helpers/strings.spec.js [hunk 9]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
 
         expect(template({obj})).toEqual('Hello Kabir! How are you?');
     });
 
+    /* lowercase */
+    it('lowercase should return lowercase value of a string param', () => {
+        expect(strings.lowercase('Hello World!')).toEqual('hello world!');
+    });
+
+    it('lowercase should return number for interger param', () => {
```
Comment (kabirbaidhya) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 10]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        expect(strings.lowercase(1234)).toEqual(1234);
+    });
+
+    it('lowercase should return empty for empty param', () => {
```
Comment (kabirbaidhya) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 11]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        expect(strings.lowercase()).toEqual('');
+    });
+
+    it('lowercase should return boolean value for boolean param', () => {
```
Comment (kabirbaidhya) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 12]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        expect(strings.lowercase(true)).toEqual(true);
+    });
+
+    it('lowercase should return null for null param', () => {
```
Comment (kabirbaidhya) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 13]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+    it('lowercase should work as expected for string param after compilation (Basic Support)', () => {
+        var template = compile('{{lowercase val}}');
+        var obj = {
+            val: 'JUST WOW!!!'
+        };
+
+        expect(template(obj)).toEqual('just wow!!!');
+    });
+
+    it('lowercase should work as expected for integer param after compilation (Basic Support)', () => {
```
Comment (kabirbaidhya) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 14]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        var template = compile('{{lowercase val}}');
+        var obj = {
+            val: 1234
+        };
+
+        expect(template(obj)).toEqual('1234');
+    });
+
+    it('lowercase should work as expected for empty param after compilation (Basic Support)', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 15]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        var template = compile('{{lowercase val}}');
+        var obj = {};
+
+        expect(template(obj)).toEqual('');
+    });
+
+    it('lowercase should work as expected for boolean param after compilation (Basic Support)', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 16]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        var template = compile('{{lowercase val}}');
+        var obj = {
+            val: true
+        };
+
+        expect(template(obj)).toEqual('true');
+    });
+
+    it('lowercase should work as expected for null param after compilation (Basic Support)', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 17]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+
+        expect(template(obj)).toEqual('');
+    });
+
+    /* uppercase */
+    it('uppercase should return uppercase value of a string param', () => {
+        expect(strings.uppercase('hello world!')).toEqual('HELLO WORLD!');
+    });
+
+    it('uppercase should return number for interger param', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 18]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        expect(strings.uppercase(1234)).toEqual(1234);
+    });
+
+    it('uppercase should return empty for empty param', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 19]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        expect(strings.uppercase()).toEqual('');
+    });
+
+    it('uppercase should return boolean value for boolean param', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 20]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        expect(strings.uppercase(true)).toEqual(true);
+    });
+
+    it('uppercase should return null for null param', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 21]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+    it('uppercase should work as expected for string param after compilation (Basic Support)', () => {
+        var template = compile('{{uppercase val}}');
+        var obj = {
+            val: 'just wow!!!'
+        };
+
+        expect(template(obj)).toEqual('JUST WOW!!!');
+    });
+
+    it('uppercase should work as expected for integer param after compilation (Basic Support)', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 22]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        var template = compile('{{uppercase val}}');
+        var obj = {
+            val: 1234
+        };
+
+        expect(template(obj)).toEqual('1234');
+    });
+
+    it('uppercase should work as expected for empty param after compilation (Basic Support)', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 23]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        var template = compile('{{uppercase val}}');
+        var obj = {};
+
+        expect(template(obj)).toEqual('');
+    });
+
+    it('uppercase should work as expected for boolean param after compilation (Basic Support)', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: tests/helpers/strings.spec.js [hunk 24]
```diff
@@ -56,42 +55,604 @@ describe('strings', () => {
...
+        var template = compile('{{uppercase val}}');
+        var obj = {
+            val: true
+        };
+
+        expect(template(obj)).toEqual('true');
+    });
+
+    it('uppercase should work as expected for null param after compilation (Basic Support)', () => {
```
Comment (sanjeevkpandit) [inline tests/helpers/strings.spec.js]: unnecessary

File: src/helpers/strings.js [hunk 25]
```diff
@@ -120,5 +119,109 @@ export default {
...
+     * @return string
+     */
+    uppercase: (param) => {
+        return isString(param) ? param.toUpperCase() : param;
+    },
+
+    /**
+     * Get the first element of a collection/array.
+     * Example usage:
+     * 		{{first ['David', 'Miller', 'Jones']}}   => 'David'
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: handlebars don't support expressions like this `{{first ['David', 'Miller', 'Jones']}}` it should be like this `{{first someArray}}`

File: src/helpers/strings.js [hunk 26]
```diff
@@ -120,5 +119,109 @@ export default {
...
+            return '';
+        }
+
+        return collection[0];
+    },
+
+    /**
+     * Get the last element of a collection/array.
+     * Example usage:
+     * 		{{last ['David', 'Miller', 'Jones']}}   => 'Jones'
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: same here

File: src/helpers/strings.js [hunk 27]
```diff
@@ -120,5 +119,109 @@ export default {
...
+            }
+        }
+
+        return resultString;
+    },
+
+    /**
+     * Join the elements of an array using a delimeter.
+     * Example usage:
+     * 	    {{join ['Hands', 'legs', 'feet'] ' & '}}   => 'Hands & legs & feet'
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: same here. BTW, you can have an example like this: 
```javascript
var someArray = ['Hands', 'legs', 'feet'];
```
```html
{{join someArray ` & '}}   => 'Hands & legs & feet'
```

File: src/helpers/strings.js [hunk 28]
```diff
@@ -120,5 +119,109 @@ export default {
...
+     *
+     * @param  array params
+     * @param  string delimeter
+     * @return string
+     */
+    join: (params = [], delimeter) => {
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: there's no use of default value `params = []`. Just keep it `params`

File: src/helpers/strings.js [hunk 29]
```diff
@@ -120,5 +119,109 @@ export default {
...
+        var resultString = '';
+
+        if (typeof delimeter === 'undefined' || delimeter === null || isObject(delimeter)) {
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: Can shorten this to: 
```
if (!delimeter || isObject(delimeter)) {
    delimeter = '';
}
```

File: src/helpers/strings.js [hunk 30]
```diff
@@ -120,5 +119,109 @@ export default {
...
+            delimeter = '';
+        }
+
+        if (params !== null) {
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]:
```javascript
if (!isArray(params)) {
    return false;
}
```

File: tests/helpers/strings.spec.js [hunk 31]
```diff
@@ -112,4 +112,134 @@ describe('strings', () => {
...
+        it('concat should return concatenation of all param values(string)', () => {
+            expect(strings.concat('hello', ' ', 'world', '!!!')).toEqual('hello world!!!');
+        });
+
+        it('concat should return concatenation of all param values(string and integer)', () => {
+            expect(strings.concat('I have got', ' ', 4, ' ', 'apples.')).toEqual('I have got 4 apples.');
+        });
+
+        it('concat should work as expected for string params after compilation (Basic Support)', () => {
+            var template = compile('{{concat first \' \' middle \' \' last}}');
```
Comment (kabirbaidhya) [inline tests/helpers/strings.spec.js]: you can use double quotes inside the string here. :)
eg: `var template = compile('{{concat first " " middle " " last}}');`

File: src/helpers/strings.js [hunk 32]
```diff
@@ -120,5 +119,114 @@ export default {
...
+     * Concat two or more strings.
+     * Example usage:
+     * 	    {{concat 'Hello' ' world' '!!!'}}   => 'Hello world!!!'
+     *
+     * @param  mixed ...params
+     * @return string
+     */
+    concat: (...params) => {
+        var resultString = '';
+        for (var i in params) {
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: How about just `return params.join('')`. DRY :wink: 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join

File: src/helpers/strings.js [hunk 33]
```diff
@@ -120,5 +119,114 @@ export default {
...
+
+        if (!delimeter || isObject(delimeter)) {
+            delimeter = '';
+        }
+
+        if (!isArray(params)) {
+            return false;
+        }
+
+        for (var i = 0; i < params.length; i++) {
```
Comment (kabirbaidhya) [inline src/helpers/strings.js]: same here. `params.join(delimeter);`

File: tests/helpers/strings.spec.js [hunk 34]
```diff
@@ -112,4 +112,134 @@ describe('strings', () => {
...
+                    'Miller',
+                    'Jones'
+                ]
+            };
+
+            expect(template(obj)).toEqual('Jones');
+        });
+    });
+
+
```
Comment (kabirbaidhya) [inline tests/helpers/strings.spec.js]: extra line :trollface:

File: tests/helpers/strings.spec.js [hunk 35]
```diff
@@ -112,4 +112,134 @@ describe('strings', () => {
...
+                s1: 'I have got ',
+                amount: 4,
+                s2: ' apples.'
+            };
+
+            expect(template(sentence)).toEqual('I have got 4 apples.');
+        });
+    });
+
+    /* join */
```
Comment (kabirbaidhya) [inline tests/helpers/strings.spec.js]: no need of this anymore as `describe('join', () => {` **describes** it better :wink:

File: src/helpers/strings.js (hunk 32 above)
Comment (sanjeevkpandit) [inline src/helpers/strings.js]: :+1:

**Comment by kabirbaidhya** [comment]
All looks good. Can you do `npm test` and paste the test results for string helpers here in the comment? Just for some chahal pahal :smile:

**Comment by sanjeevkpandit** [comment]
```
strings
    excerpt
      ✓ should extract all the characters from a string if it is less than 50 characters by default
      ✓ should extract 50 characters from a string if it has more than 50 characters by default
      ✓ should extract provided number of characters from a string
      ✓ should extract all the characters from a string if the provided number of characters to be extracted is more than the number of characters
      ✓ should return the string if the length parameter is not a number
    sanitize
      ✓ should return a normal string as dash case
      ✓ should return a string with special characters as dash case without special characters
    capitalizeFirst
      ✓ should capitalize the first letter of a string
      ✓ should return the param if it is not a string
    capitalizeEach
      ✓ should capitalize the first letter of a string
      ✓ should return the param if it is not a string
    sprintf
      ✓ function should work as expected (basic support)
      ✓ should work as expected (Basic support)
      ✓ should work as expected after compilation (Basic support)
      ✓ should work as expected after compilation (C-style sprintf)
      ✓ should work as expected after compilation (C-style sprintf) with arbitrary number of parameters
      ✓ should work as expected after compilation if an object is passed dynamically
    lowercase
      ✓ lowercase should return lowercase value of a string param
      ✓ lowercase should should work as expected after compilation (Basic Support)
    uppercase
      ✓ uppercase should return uppercase value of a string param
      ✓ uppercase should work as expected after compilation (Basic Support)
    first
      ✓ first should return first element of an array(string)
      ✓ first should work as expected after compilation (Basic Supprt)
    last
      ✓ last should return last element of an array(string)
      ✓ last should work as expected after compilation (Basic Supprt)
    concat
      ✓ concat should return concatenation of all param values(string)
      ✓ concat should return concatenation of all param values(string and integer)
      ✓ concat should work as expected for string params after compilation (Basic Support)
      ✓ concat should work as expected for string and integer params after compilation (Basic Support)
    join
      ✓ join should join the values of array of strings using the delimeter provided
      ✓ join should return concatenation of elements of array using empty string if no delimeter provided
      ✓ join should work as expected for array of strings after compilation (Basic Support)
      ✓ join should work as expected for some array and no delimeter after compilation (Basic Support)

```

**Comment by kabirbaidhya** [comment]
Awesome :100:
